### PR TITLE
update pool states

### PR DESCRIPTION
**Pull request by trgill**
Update the PoolState to mirror the DM state of a pool.  DM pools run in 4 possible modes (PM_OUT_OF_METADATA_SPACE and PM_READ_ONLY are reported as the same to user space).

DM mode enum:

```
        PM_WRITE,               /* metadata may be changed */
        PM_OUT_OF_DATA_SPACE,   /* metadata may be changed, though data may not be allocated */
        /*
         * Like READ_ONLY, except may switch back to WRITE on metadata resize. Reported as READ_ONLY.
         */
        PM_OUT_OF_METADATA_SPACE,
        PM_READ_ONLY,           /* metadata may not be changed */

        PM_FAIL,  
``` 
We add two additional modes to the pool for Initializing and Stopping.

Add publishing the mode on dbus with signals to indicate changes.

PoolExtendState added to allow an API user to track any problems with extending the data or meta data devices.

Publish the FreeSpaceState on dbus and associated change signal.

### CODE REVIEW

**Comment by agrover** [comment]
I think this code looks good, the main piece I'd like to discuss is giving a little more structure by using a hierarchy: states and substates. i.e. Stopped/Started (or Working/NotWorking, or some other combination of exhaustive states) that encompass the high-level piece of: is the pool accessible or not? Then, within each of those, there can be a substate: stopped/initializing, stopped/incomplete, started/good, started/degraded, etc. I'm sure we're going to be adding more (sub)states as time goes on, and putting them all in a flat list as we add them could make answering the basic question of "is this pool up?" tricky to do.

@trgill and others, what do you think? This is something we either need to figure out before 1.0 if we freeze API, OR, we could go with this for 1.0, but explicitly *not* declare the API was stable, and have another dev cycle to discuss the issue.

So to sum up:
1. what do you think of the states/substates idea? If you don't :heart: it maybe we should:
1. Consider not declaring API stability for 1.0 so the team has more time to discuss. (This would make me likely to approve this PR since it is a big step up from what we have now)

**Comment by mulkieran** [comment]
I'm all about not declaring API stability.

**Comment by agrover** [comment]
OK, let's not stabilize API in 1.0. That makes things a little easier :smiley:. I'll have some normal review comments with that in mind shortly, thanks.

File: src/engine/types.rs [hunk 1]
```diff
@@ -22,17 +22,62 @@ pub enum RenameAction {
...
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum PoolExtendState {
+    Initializing,
+    Good,
+    Failed {
+        meta_failed: bool,
+        data_failed: bool,
```
Comment (agrover) [inline src/engine/types.rs]: Consider MetaFailed, DataFailed, and MetaAndDataFailed states here? There aren't that many possibilities, and this would make it impossible to be in a Failed(false, false) state. (this could either be at the toplevel of the enum, or there could be a PoolExtendFailureState within Failed that was Meta, Data, or Both.)

File: src/engine/strat_engine/thinpool/thinpool.rs [hunk 2]
```diff
@@ -562,10 +574,11 @@ impl ThinPool {
 
                 self.thin_pool.set_low_water_mark(get_dm(), lowater)?;
                 self.resume()?;
+                self.set_extend_state(data_extend_failed, meta_extend_failed);
```
[the remark this answers is not in the document]
Comment (trgill) [reply src/engine/strat_engine/thinpool/thinpool.rs]: There is a way out of Failed...  in set_extend_state()

    fn set_extend_state(&mut self, data_extend_failed: bool, meta_extend_failed: bool) {
        let mut new_state = PoolExtendState::Good;

We set it to good if an extension hasn't failed.

I was thinking that each iteration of check() will trigger the same extension attempt, right?  So, if the meta extend failed in one iteration - it will either fail again or something worked and it was extended.

I realize that might not be a good feature of check - that we would keep attempting the extends.

Am I missing something?

[the remark this answers is not in the document]
Comment (agrover) [reply src/engine/strat_engine/thinpool/thinpool.rs]: ah ok, no that is true! haha thanks.

File: src/engine/types.rs (hunk 1 above)
Comment (trgill) [reply src/engine/types.rs]: Ok - thanks.   Updating now.

**Comment by agrover** [comment]
oh no, you rebased, didn't you? Oh well.

Anyway, code looks good. Thanks.
